primitives/src/crh/bowe_hopwood/mod.rs

```diff
@@ -17,14 +18,87 @@ pub const CHUNK_SIZE: usize = 3;
 #[serde(bound(deserialize = "G: Group"))]
 pub struct BoweHopwoodPedersenParameters<G: Group> {
     pub generators: Vec<Vec<G>>,
+    pub maximum_num_chunks_in_segment: usize,
 }
 
-pub struct BoweHopwoodPedersenCRH<G: Group, W: PedersenWindow> {
+pub struct BoweHopwoodPedersenCRH<G: Group, W: PedersenWindow, const N: usize> {
     group: PhantomData<G>,
     window: PhantomData<W>,
 }
 
-impl<G: Group, W: PedersenWindow> BoweHopwoodPedersenCRH<G, W> {
+impl<G: Group, W: PedersenWindow, const N: usize> BoweHopwoodPedersenCRH<G, W, N> {
+
+    pub fn check_setup_preconditions(maximum_num_chunks_in_segment: usize) -> Result<(), Error> {
+        assert_eq!(CHUNK_SIZE, 3);
+
+        if Self::INPUT_SIZE_BITS > W::WINDOW_SIZE * W::NUM_WINDOWS * CHUNK_SIZE {
+            return Err(Box::new(CryptoError::Other(
+                format!(
+                    "window params {:?}x{:?}x{} are incorrect for input length {:?}",
+                    W::WINDOW_SIZE,
+                    W::NUM_WINDOWS,
+                    CHUNK_SIZE,
+                    N,
+                )
+                .to_owned(),
+            )));
+        }
+
+        if W::WINDOW_SIZE > maximum_num_chunks_in_segment {
+            return Err(format!(
+                "Bowe-Hopwood hash must have a window size resulting in scalars < (p-1)/2, \
+                 maximum segment size is {}",
+                 maximum_num_chunks_in_segment
+            )
+            .into());
+        }
+
+        Ok(())
+    }
+
+    pub fn check_preconditions(params: &BoweHopwoodPedersenParameters<G>) -> Result<(), Error> {
+
+        Self::check_setup_preconditions(params.maximum_num_chunks_in_segment)?;
+
+        if params.generators.len() != W::NUM_WINDOWS {
+            Err(Box::new(CryptoError::Other(
+                format!(
+                    "Incorrect pp of size {:?} for window params {:?}x{:?}x{}",
+                    params.generators.len(),
+                    W::WINDOW_SIZE,
+                    W::NUM_WINDOWS,
+                    CHUNK_SIZE
+                )
+                .to_owned(),
+            )))?
+        }
+
+        for generators in params.generators.iter() {
+            if generators.len() != W::WINDOW_SIZE {
+                Err(Box::new(CryptoError::Other(format!(
+                    "Number of generators: {} not enough for the selected window size: {}",
+                    params.generators.len(),
+                    W::WINDOW_SIZE
+                ))))?
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Zcash protocol specification - section 5.4.1.7
+    fn calculate_num_chunks_in_segment<F: PrimeField>() -> usize {
+        let upper_limit: BigUint = F::modulus_minus_one_div_two().into();
+        let mut c = 0;
+        let mut range = BigUint::from(2usize);
+        while range < upper_limit {
+            range <<= 4;
+            c += 1;
+        }
+
+        c
```

Comment:
Shouldn't this return c-1? According to my understanding c should be the highest integer such that for which `range < upper_limit`, but in this case we return the first value `c` for which the condition does not hold. Indeed, the maximum `WINDOW_SIZE` for tweedle should be 63 (as in ZCash specs) but the tests run without issues for `WINDOW_SIZE=64`

Reply:
You are right, nice catch !